### PR TITLE
Feat/node group template node info

### PR DESCRIPTION
**Pull request by visdmin**
#### What type of PR is this?
/kind feature

#### What this PR does:
**Add support for node groups with `MinSize: 0`** 

- Adds additional upcloud-go-api components required for fetching server plans:
    - `service/cloud.go`
    - `plan.go` `price.go` `timezone.go` `zone.go` (required by `service/cloud.go`)
- Fetch and store node group CPU count and memory amount from upc api
    Fetch all plans (api does not seem to have filtering support) and find the matching plan details for the group plan.
- Store labels and taints for `upCloudNodeGroup` there are used by the node group `TemplateNodeInfo` method
- Implements node group `TemplateNodeInfo` method
- Changes to constants
    - change `nodeGroupMinSize` `1` -> `0`
    - add `nodeMaxPods` `110`
- Unit tests:
    - Add test case that validates the `NodeInfo` returned by `TemplateNodeInfo` of empty node group
    - Add test case that validates that in case only single node group is defined it cannot have `MinSize: 0`
    - Change tests to include node group plan
    - Change tests to allow `SupportScaleToZero: true`
    - Change test to verify that node group with `MinSize: 0` is valid and allowed when at least one other node group exists with `MinSize >= 1`.

#### Why we need it:
Enables great **cost** and environmental saving opportunities for UKS users. 

#### Special notes for your reviewer:
- PR author has very limited experience with Go lang, so please do not get frustrated with the coding and naming convention mistakes, just let me know what changes would be needed.  
- Node group with `MinSize: 0` is only allowed when at least one node group with `MinSize >= 1` exists
    This is because the UpC cluster auto-scaler is not running inside control plane.
- PR author does not have access to information about node templates used by UKS clusters
    `NodeInfo` returned by `TemplateNodeInfo` is very likely incomplete.
- Node group `TemplateNodeInfo` currently returns `cloudprovider.ErrNotImplemented` when group size is `> 0`, this may be against conventions ? (but preserves old behaviour for node groups that have node)  
- PR author is not sure how `NodeGroupSpec.SupportScaleToZero` should treated:
    This property was previously set to `false` by using constant `nodeGroupMinSize (1)` this constant has been changed to `0` and therefore the `NodeGroupSpec.SupportScaleToZero` is always set to `true`. Technically UKS node groups support scaling to 0. Author is just bit confused if the `NodeGroupSpec.SupportScaleToZero` should be set to `false` for node groups that have `MinSize >= 1`
- New upcloud-go-api components have been added running: `vendor.sh github.com/upcloudltd/upcloud-go-api/v6/ v6.5.0`
    

#### Does this PR introduce a user-facing change?
```release-note
### Enables scaling down node groups to zero
Setting node group minimum size to 0 requires that least one node group has to have minimum size >= 1 
```

### CODE REVIEW

File: cluster-autoscaler/cloudprovider/upcloud/pkg/github.com/upcloudltd/upcloud-go-api/v6/upcloud/zone.go [hunk 1]
```diff
@@ -0,0 +1,35 @@
+package upcloud
+
+import "encoding/json"
+
+// Zones represents a /zone response
+type Zones struct {
```
Comment (mclang) [inline cluster-autoscaler/cloudprovider/upcloud/pkg/github.com/upcloudltd/upcloud-go-api/v6/upcloud/zone.go:6]:
```suggestion
type AvailabilityZones struct {
```

File: cluster-autoscaler/cloudprovider/upcloud/pkg/github.com/upcloudltd/upcloud-go-api/v6/upcloud/zone.go [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+		return err
+	}
+
+	s.Zones = v.Zones.Zones
+
+	return nil
+}
+
+// Zone represents a zone
+type Zone struct {
```
Comment (mclang) [inline cluster-autoscaler/cloudprovider/upcloud/pkg/github.com/upcloudltd/upcloud-go-api/v6/upcloud/zone.go:31]:
```suggestion
type AvailabilityZone struct {
```

File: cluster-autoscaler/cloudprovider/upcloud/pkg/github.com/upcloudltd/upcloud-go-api/v6/upcloud/zone.go [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+import "encoding/json"
+
+// Zones represents a /zone response
+type Zones struct {
+	Zones []Zone `json:"zone"`
+}
+
+// UnmarshalJSON is a custom unmarshaller that deals with
+// deeply embedded values.
+func (s *Zones) UnmarshalJSON(b []byte) error {
```
Comment (mclang) [inline cluster-autoscaler/cloudprovider/upcloud/pkg/github.com/upcloudltd/upcloud-go-api/v6/upcloud/zone.go:12]:
```suggestion
func (s *AvailabilityZones) UnmarshalJSON(b []byte) error {
```

File: cluster-autoscaler/cloudprovider/upcloud/upcloud_manager_test.go [hunk 4]
```diff
@@ -74,18 +74,39 @@ func TestManager(t *testing.T) {
...
-			NodeGroupSpecs: []string{"1:2:one", "11:20:two"},
+			NodeGroupSpecs: []string{"1:2:one", "11:20:two", "0:10:three"},
 		},
 	)
 	require.NoError(t, err)
 	require.Equal(t, upCfg.ClusterID, m.clusterID.String())
-	require.Equal(t, dynamic.NodeGroupSpec{Name: "one", MinSize: 1, MaxSize: 2}, m.nodeGroupSpecs["one"])
-	require.Equal(t, dynamic.NodeGroupSpec{Name: "two", MinSize: 11, MaxSize: 20}, m.nodeGroupSpecs["two"])
+	require.Equal(t, dynamic.NodeGroupSpec{Name: "one", MinSize: 1, MaxSize: 2, SupportScaleToZero: true}, m.nodeGroupSpecs["one"])
+	require.Equal(t, dynamic.NodeGroupSpec{Name: "two", MinSize: 11, MaxSize: 20, SupportScaleToZero: true}, m.nodeGroupSpecs["two"])
```
Comment (mclang) [inline cluster-autoscaler/cloudprovider/upcloud/upcloud_manager_test.go:83]: Should `SupportScaleToZero` be true|false in these cases where `MinSize > 0` :raised_eyebrow:

**Comment by visdmin** [comment]
Noihin `cluster-autoscaler/cloudprovider/upcloud/pkg/github.com/upcloudltd/upcloud-go-api/v6/*` polun tiedostoihin ei voi tehdä muutoksia, kyseessä on UpCloudin tekemä Go API clientti.

En tiedä miksi, mutta UpCloud on päättänyt lisätä se tähän projektiin suoraan lähdekoodina, joten samaa tapaa on seuratttu. 

https://pkg.go.dev/github.com/UpCloudLtd/upcloud-go-api?utm_source=godoc
https://github.com/UpCloudLtd/upcloud-go-api?tab=readme-ov-file

File: cluster-autoscaler/cloudprovider/upcloud/pkg/github.com/upcloudltd/upcloud-go-api/v6/upcloud/zone.go (hunk 1 above)
Comment (visdmin) [reply cluster-autoscaler/cloudprovider/upcloud/pkg/github.com/upcloudltd/upcloud-go-api/v6/upcloud/zone.go:6]: Noihin `cluster-autoscaler/cloudprovider/upcloud/pkg/github.com/upcloudltd/upcloud-go-api/v6/*` polun tiedostoihin ei voi tehdä muutoksia, kyseessä on UpCloudin tekemä Go API clientti.

En tiedä miksi, mutta UpCloud on päättänyt lisätä se tähän projektiin suoraan lähdekoodina, joten samaa tapaa on seuratttu. 

https://pkg.go.dev/github.com/UpCloudLtd/upcloud-go-api?utm_source=godoc
https://github.com/UpCloudLtd/upcloud-go-api?tab=readme-ov-file

File: cluster-autoscaler/cloudprovider/upcloud/pkg/github.com/upcloudltd/upcloud-go-api/v6/upcloud/zone.go (hunk 2 above)
Comment (visdmin) [reply cluster-autoscaler/cloudprovider/upcloud/pkg/github.com/upcloudltd/upcloud-go-api/v6/upcloud/zone.go:31]: Noihin `cluster-autoscaler/cloudprovider/upcloud/pkg/github.com/upcloudltd/upcloud-go-api/v6/*` polun tiedostoihin ei voi tehdä muutoksia, kyseessä on UpCloudin tekemä Go API clientti.

En tiedä miksi, mutta UpCloud on päättänyt lisätä se tähän projektiin suoraan lähdekoodina, joten samaa tapaa on seuratttu. 

https://pkg.go.dev/github.com/UpCloudLtd/upcloud-go-api?utm_source=godoc
https://github.com/UpCloudLtd/upcloud-go-api?tab=readme-ov-file

File: cluster-autoscaler/cloudprovider/upcloud/pkg/github.com/upcloudltd/upcloud-go-api/v6/upcloud/zone.go (hunk 3 above)
Comment (visdmin) [reply cluster-autoscaler/cloudprovider/upcloud/pkg/github.com/upcloudltd/upcloud-go-api/v6/upcloud/zone.go:12]: Noihin `cluster-autoscaler/cloudprovider/upcloud/pkg/github.com/upcloudltd/upcloud-go-api/v6/*` polun tiedostoihin ei voi tehdä muutoksia, kyseessä on UpCloudin tekemä Go API clientti.

En tiedä miksi, mutta UpCloud on päättänyt lisätä se tähän projektiin suoraan lähdekoodina, joten samaa tapaa on seuratttu. 

https://pkg.go.dev/github.com/UpCloudLtd/upcloud-go-api?utm_source=godoc
https://github.com/UpCloudLtd/upcloud-go-api?tab=readme-ov-file

File: cluster-autoscaler/cloudprovider/upcloud/upcloud_manager_test.go (hunk 4 above)
Comment (visdmin) [reply cluster-autoscaler/cloudprovider/upcloud/upcloud_manager_test.go:83]: Ei tietoa, tästä on UpC:lle laitettu kysymys sinnepäin menevässä PRssä.

https://pkg.go.dev/k8s.io/autoscaler/cluster-autoscaler/config/dynamic#NodeGroupSpec

```go
type NodeGroupSpec struct {
    // The name of the autoscaling target
    Name [string](https://pkg.go.dev/builtin#string) `json:"name"`
    // Min size of the autoscaling target
    MinSize [int](https://pkg.go.dev/builtin#int) `json:"minSize"`
    // Max size of the autoscaling target
    MaxSize [int](https://pkg.go.dev/builtin#int) `json:"maxSize"`
    // Specifies whether this node group can scale to zero nodes.
    SupportScaleToZero [bool](https://pkg.go.dev/builtin#bool)
}
```

Tässä on kaksi vaihtoehto miten tuon `SupportScaleToZero` voi ymmärtää:
- Tukeeko kyseinen node group scalaamista 0 asti (mutta MinSize voi estää tämän)
- Halutaanko että kyseinen node group voidaan scaalata 0 asti (vaikka MinSize > 0)


Joka tapauksessa tämä on hankala homma koska tuo NodeGroupSpec luodaan käyttäjältä saadusta configuraatiosta tälläistä functiota käyttäen:

([kutsutaan täällä](https://github.com/kihoteam/autoscaler/pull/1/files#r1959720494))
```go
// SpecFromString parses a node group spec represented in the form of `<minSize>:<maxSize>:<name>` and produces a node group spec object
func SpecFromString(value string, SupportScaleToZero bool) (*NodeGroupSpec, error) {
	tokens := strings.SplitN(value, ":", 3)
	if len(tokens) != 3 {
		return nil, fmt.Errorf("wrong nodes configuration: %s", value)
	}

	spec := NodeGroupSpec{SupportScaleToZero: SupportScaleToZero}
	if size, err := strconv.Atoi(tokens[0]); err == nil {

		spec.MinSize = size
	} else {
		return nil, fmt.Errorf("failed to set min size: %s, expected integer", tokens[0])
	}

	if size, err := strconv.Atoi(tokens[1]); err == nil {
		spec.MaxSize = size
	} else {
		return nil, fmt.Errorf("failed to set max size: %s, expected integer", tokens[1])
	}

	spec.Name = tokens[2]

	if err := spec.Validate(); err != nil {
		return nil, fmt.Errorf("invalid node group spec: %v", err)
	}

	return &spec, nil
}

// Validate produces an error if there's an invalid field in the node group spec
func (s NodeGroupSpec) Validate() error {
	if s.SupportScaleToZero {
		if s.MinSize < 0 {
			return fmt.Errorf("min size must be >= 0")
		}
	} else {
		if s.MinSize <= 0 {
			return fmt.Errorf("min size must be >= 1")
		}
	}
	if s.MaxSize < s.MinSize {
		return fmt.Errorf("max size must be greater or equal to min size")
	}
	if s.Name == "" {
		return fmt.Errorf("name must not be blank")
	}
	return nil
}
```

Eli tuo `SupportScaleToZero` pitää antaa funcioon ennekun tiedetään mikä `MinSize` on.
Functio ei ole UpCloudin vaan osa autoscaler repoa jo valmiiksi, joten olettaisin että tuo `SupportScaleToZero` pitää ymmärtää niin että TUKEEKO node group ylipäätään tuota vai ei.

File: cluster-autoscaler/cloudprovider/upcloud/upcloud_manager.go [hunk 5]
```diff
@@ -126,6 +135,7 @@ func nodeGroupSpecsFromDiscoveryOptions(do *cloudprovider.NodeGroupDiscoveryOpti
 	if do == nil || len(do.NodeGroupSpecs) == 0 {
 		return specs, nil
 	}
+
 	for _, spec := range do.NodeGroupSpecs {
 		s, err := dynamic.SpecFromString(spec, supportScaleToZero)
```
Comment (visdmin) [inline cluster-autoscaler/cloudprovider/upcloud/upcloud_manager.go:140]: Täällä kutsutaan tuota `dynamic.SpecFromString`

**Review by vilnuu — APPROVED** [review]
